server/src/main/java/org/elasticsearch/threadpool/ScalingExecutorBuilder.java

```diff
@@ -173,4 +193,50 @@ static class ScalingExecutorSettings extends ExecutorBuilder.ExecutorSettings {
         }
     }
 
+    private static class KeepAliveTimeValidator implements Setting.Validator<TimeValue> {
+        private final Setting<Integer> coreSetting;
+        private final Setting<Integer> maxSetting;
+
+        KeepAliveTimeValidator(Setting<Integer> coreSetting, Setting<Integer> maxSetting) {
+            this.coreSetting = coreSetting;
+            this.maxSetting = maxSetting;
+        }
+
+        @Override
+        public void validate(TimeValue value) {
+            if (value.duration() < 0) {
+                throw new IllegalArgumentException("Invalid negative value for [" + prefix() + "keep_alive].");
+            }
+        }
+
+        @Override
+        public void validate(TimeValue value, Map<Setting<?>, Object> settings) {
+            if (value.duration() > 0) {
+                return;
+            }
+            if (settings.get(coreSetting).equals(0) && settings.get(maxSetting).equals(1)) {
+                throw new IllegalArgumentException(
+                    "["
+                        + prefix()
+                        + "keep_alive] must be greater than 0 when ["
+                        + coreSetting.getKey()
+                        + "] is [0] and ["
+                        + coreSetting.getKey()
+                        + "] is [1]."
+                );
```

Comment:
Hmm I'm suspicious about this, the bug exists regardless of keepalive time doesn't it? It's just a little harder to trigger in that case.

Reply:
`newSingleScalingToZero` (which is used in this case) uses `allowCoreThreadTimeOut=true`, this requires a non-zero keep alive value. this doesn't use any of the ES scaling magic to make the pool scale beyond core size despite using an unbounded queue.

Reply:
ok, that wasn't needed in the end, reverted